compiler/z/codegen/OMRTreeEvaluator.cpp

```diff
@@ -14117,6 +14117,168 @@ bool isSettingOp(uint8_t byteValue, TR::InstOpCode::Mnemonic SI_opcode)
    }
 
 
+
+TR::Register *
+OMR::Z::TreeEvaluator::bitpermuteEvaluator(TR::Node *node, TR::CodeGenerator *cg)
+   {
+
+   TR_ASSERT(node->getNumChildren() == 3, "Wrong number of children in bitpermuteEvaluator");
+   TR::Node *value = node->getChild(0);
+   TR::Node *addr = node->getChild(1);
+   TR::Node *length = node->getChild(2);
+
+   bool nodeIs64Bit = node->getSize() == 8;
+   auto valueReg = cg->evaluate(value); // this will either be a constant or a load
+   auto addrReg = cg->evaluate(addr);   // this will either be a constant or a load
+
+   TR::Register *tmpReg = cg->allocateRegister(TR_GPR);
+
+   //Zero result reg
+   TR::Register *resultReg = cg->allocateRegister(TR_GPR);
+   generateRRInstruction(cg, TR::InstOpCode::getXORRegOpCode(), node, resultReg, resultReg);
+
+   if (length->getOpCode().isLoadConst())
+      {
+      // Manage the constant length case
+      uintptrj_t arrayLen = length->getIntegerNodeValue<uintptrj_t>();
+      for (uintptrj_t x = 0; x < arrayLen; ++x)
+         {
+         // zero tmpReg
+         generateRRInstruction(cg, TR::InstOpCode::getXORRegOpCode(), node, tmpReg, tmpReg);
+
+         //load the shift amount into tmpReg
+         //TODO: Check to see if we need sourceMR->stopUsingMemRefRegister(cg)
+         TR::MemoryReference *sourceMR = generateS390MemoryReference(addrReg, x, cg, NULL);
+         generateRXYInstruction(cg, TR::InstOpCode::LB, node, tmpReg, sourceMR);
+
+         if (!(node->getDataType() == TR::Int32 || node->getDataType() == TR::Int16 || node->getDataType() == TR::Int8 || node->getDataType() == TR::Int64))
+            {
+            TR_ASSERT(0, " Unsupported DataType for bitPermute op");
+            }
+         if (node->getDataType() != TR::Int64)
+            {
+            // Java Spec defines the shift amount for integer shifts to be masked with
```

Comment:
I don't think we should mention the Java Spec here. It is just the way our IL is modeled at the moment.

Reply:
So I left this comment in here on purpose. I wasn't sure if this is something we should even be doing at all. So I thought it's best to leave this here for now and decide on a solution before this is merged. I also opened an issue discussing this here: https://github.com/eclipse/omr/issues/2320, but no final verdict has been reached on whether these masks should be removed or not.

For now, should we mimic how the rest of the JIT handles this case? We can save two instructions in this sequence if we don't do this masking.

Reply:
I don't think so, the IL defines the following:

https://github.com/eclipse/omr/blob/97107f4d8b0d076e8f848d11e9fae8dc521e0b94/compiler/il/ILOpCodeProperties.hpp#L11708-L11711

This means we are free to do whatever if the shift is out of bounds. This implies we don't need to mask the value via `NILL`.

Reply:
Okay so my initial interpretation of that comment was since the rest of the VM tries to mimic hardware behaviour and there are multiple instances inside OMR where we do this masking, it should be done here as well for consistency. However, upon thinking about this more, I think you are right and it shouldn't be the codgen's responsibility to mask the shift amount. The IL specification rules should be obeyed by whoever generated it in the first place. 

This is addressed in: bc59479

This also saves us two instructions on the 32-bit side.